### PR TITLE
Enabling large tensor support for binary broadcast operators

### PR DESCRIPTION
**Pull request by ChaiBapchya**
## Description ##
In continuation of #16714 
Add tests
- arctan2
- hypot

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] tests/nightly/test_large_array.py
- [x] src/operator/tensor/elemwise_binary_broadcast_op.h

### CODE REVIEW

**Comment by access2rohit** [comment]
LGTM ! Also paste the output for tests run.
Can we also do large vector tests too for arctan2 ? I assume hypt is hypotenuse so maybe that requires 2D matrix

**Comment by ChaiBapchya** [comment]
Build flag
```
python -c "from mxnet.runtime import feature_list; print(feature_list())"
[✖ CUDA, ✖ CUDNN, ✖ NCCL, ✖ CUDA_RTC, ✖ TENSORRT, ✔ CPU_SSE, ✔ CPU_SSE2, ✔ CPU_SSE3, ✔ CPU_SSE4_1, ✔ CPU_SSE4_2, ✖ CPU_SSE4A, ✔ CPU_AVX, ✖ CPU_AVX2, ✔ OPENMP, ✖ SSE, ✔ F16C, ✖ JEMALLOC, ✔ BLAS_OPEN, ✖ BLAS_ATLAS, ✖ BLAS_MKL, ✖ BLAS_APPLE, ✔ LAPACK, ✔ MKLDNN, ✔ OPENCV, ✖ CAFFE, ✖ PROFILER, ✔ DIST_KVSTORE, ✖ CXX14, ✔ INT64_TENSOR_SIZE, ✔ SIGNAL_HANDLER, ✖ DEBUG, ✖ TVM_OP]
```
Output of test run
```
ubuntu@ip-172-31-27-86:~/workspace/incubator-mxnet$ nosetests tests/nightly/test_large_array.py:test_binary_broadcast
.
----------------------------------------------------------------------
Ran 1 test in 39.432s

OK
```

**Comment by ChaiBapchya** [comment]
> LGTM ! Also paste the output for tests run.
> Can we also do large vector tests too for arctan2 ? I assume hypt is hypotenuse so maybe that requires 2D matrix

hypot can work for vector too (it takes 2 values as input which basically act as 2 sides of the triangle) - shape of the 2 values has to be the same / broadcastable

**Comment by ChaiBapchya** [comment]
Vector test
```
nosetests tests/nightly/test_large_vector.py:test_binary_broadcast
.
----------------------------------------------------------------------
Ran 1 test in 20.125s

OK
```

File: src/operator/tensor/elemwise_binary_broadcast_op.h [hunk 1]
```diff
@@ -55,10 +55,10 @@ inline bool BinaryBroadcastShape(const nnvm::NodeAttrs& attrs,
     return shape_is_known(lhs);
   }
   mxnet::TShape out(std::max(lhs.ndim(), rhs.ndim()), -1);
-  const int bl = out.ndim() - lhs.ndim();
-  const int br = out.ndim() - rhs.ndim();
-  for (int i = 0; i < out.ndim(); ++i) {
-    int l = 1, r = 1;
+  const dim_t bl = out.ndim() - lhs.ndim();
```
Comment (apeforest) [inline src/operator/tensor/elemwise_binary_broadcast_op.h]: It's a little bit strange here. ndim() should always return an `int`. Why would changing to `dim_t` fix anything? Can you actually print out the values of out, lhs and i here to see if there is indeed an integer overflow.

Comment (access2rohit) [reply src/operator/tensor/elemwise_binary_broadcast_op.h]: @apeforest : when you run this DGL [triaining](https://github.com/zheng-da/dgl-1/blob/lt_test/examples/mxnet/sampling/train.py) script you get the following error:
```
Traceback (most recent call last):
  File "examples/mxnet/sampling/train.py", line 107, in <module>
    main(args)
  File "examples/mxnet/sampling/train.py", line 69, in main
    graphsage_cv_train(g, ctx, args, n_classes, train_nid, test_nid, n_test_samples, False)
  File "/home/ubuntu/dgl/examples/mxnet/sampling/graphsage_cv.py", line 212, in graphsage_cv_train
    lambda node : {'preprocess': node.data['preprocess'] * node.data['norm']})
  File "/home/ubuntu/dgl/python/dgl/graph.py", line 2747, in update_all
    Runtime.run(prog)
  File "/home/ubuntu/dgl/python/dgl/runtime/runtime.py", line 11, in run
    exe.run()
  File "/home/ubuntu/dgl/python/dgl/runtime/ir/executor.py", line 129, in run
    udf_ret = fn_data(node_data)
  File "/home/ubuntu/dgl/python/dgl/runtime/scheduler.py", line 726, in _afunc_wrapper
    return apply_func(nbatch)
  File "/home/ubuntu/dgl/examples/mxnet/sampling/graphsage_cv.py", line 212, in <lambda>
    lambda node : {'preprocess': node.data['preprocess'] * node.data['norm']})
  File "/home/ubuntu/incubator-mxnet/python/mxnet/ndarray/ndarray.py", line 303, in __mul__
    return multiply(self, other)
  File "/home/ubuntu/incubator-mxnet/python/mxnet/ndarray/ndarray.py", line 3648, in multiply
    None)
  File "/home/ubuntu/incubator-mxnet/python/mxnet/ndarray/ndarray.py", line 3461, in _ufunc_helper
    return fn_array(lhs, rhs)
  File "<string>", line 52, in broadcast_mul
  File "/home/ubuntu/incubator-mxnet/python/mxnet/_ctypes/ndarray.py", line 107, in _imperative_invoke
    ctypes.byref(out_stypes)))
  File "/home/ubuntu/incubator-mxnet/python/mxnet/base.py", line 255, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
**mxnet.base.MXNetError: [23:32:52] include/mxnet/./tuple.h:396: Check failed: dim_size >= -1 (-1294967296 vs. -1) : shape dim size must be >= -1, while received -1294967296**
Stack trace:
  [bt] (0) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x43) [0x7fdd29ce9f13]
  [bt] (1) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::dim_size_is_known(long)+0x1a5) [0x7fdd29d51575]
  [bt] (2) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::op::BinaryBroadcastShape(nnvm::NodeAttrs const&, std::vector<mxnet::TShape, std::allocator<mxnet::TShape> >*, std::vector<mxnet::TShape, std::allocator<mxnet::TShape> >*)+0x2f0) [0x7fdd29ffc810]
  [bt] (3) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::imperative::SetShapeType(mxnet::Context const&, nnvm::NodeAttrs const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, mxnet::DispatchMode*)+0x140e) [0x7fdd2d1e22de]
  [bt] (4) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::Imperative::Invoke(mxnet::Context const&, nnvm::NodeAttrs const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&)+0xe6) [0x7fdd2d1e6c16]
  [bt] (5) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(MXImperativeInvokeImpl(void*, int, void**, int*, void***, int, char const**, char const**)+0x543) [0x7fdd2d9ea543]
  [bt] (6) /home/ubuntu/incubator-mxnet/python/mxnet/../../lib/libmxnet.so(MXImperativeInvokeEx+0x8b) [0x7fdd2d9ebe1b]
  [bt] (7) /usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_call_unix64+0x4c) [0x7fdd325f6dae]
  [bt] (8) /usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_call+0x22f) [0x7fdd325f671f]
 
```

AS you can see from the highlighted code the ndim is the parameter that is overflowing.
On a second note, in the mxnet code base we have a typedef for dimensions which is dim_t: https://github.com/apache/incubator-mxnet/blob/60d74bc948869588c2f143fd3d55231859dc979f/include/mxnet/c_api.h#L62
IMO we should use dim_t where ever possible and remove int, keeping the codebase consistent.

**Comment by access2rohit** [comment]
> Vector test
> 
> ```
> nosetests tests/nightly/test_large_vector.py:test_binary_broadcast
> .
> ----------------------------------------------------------------------
> Ran 1 test in 20.125s
> 
> OK
> ```

Did also you run the whole suite for tensor and vector ?

File: src/operator/tensor/elemwise_binary_broadcast_op.h (hunk 1 above)
Comment (ChaiBapchya) [reply src/operator/tensor/elemwise_binary_broadcast_op.h]: Yes. You're right. We don't need this change. Reverted that commit. Binary broadcast for int64 works as is. This PR will just add the test cases now.

Comment (ChaiBapchya) [reply src/operator/tensor/elemwise_binary_broadcast_op.h]: Verified this by reverting the commit, rebase master, test
```
python3 -m nose tests/nightly/test_large_array.py:test_binary_broadcast
.
----------------------------------------------------------------------
Ran 1 test in 37.253s

OK
```

**Comment by ChaiBapchya** [comment]
```
python -m nose tests/nightly/test_large_array.py:test_tvm_add
.
----------------------------------------------------------------------
Ran 1 test in 10.660s

OK
```

**Comment by ChaiBapchya** [comment]
> > Vector test
> > ```
> > nosetests tests/nightly/test_large_vector.py:test_binary_broadcast
> > .
> > ----------------------------------------------------------------------
> > Ran 1 test in 20.125s
> > 
> > OK
> > ```
> 
> Did also you run the whole suite for tensor and vector ?

Yes. As noted previously, on 480GB CPU test_large_array needs to be run twice (first half tests followed by second half tests approx) due to memory free-ing issue.

**Pull request merged**